util/circular_buffer.c

```diff
@@ -141,105 +147,183 @@ int32_t cb_remove(struct circular_buffer *desc)
-		written += writting;
-		desc->write_ptr += writting;
-		if (desc->write_ptr - desc->buff == desc->size)
-			desc->write_ptr = desc->buff;
-		to_write -= writting;
+	if (is_read) {
+		ret = cb_size(desc, &available_size);
+		if (ret == -EOVERRUN) {
+			/* Update read index */
+			desc->read.spin_count = desc->write.spin_count - 1;
+			desc->read.idx = desc->write.idx;
+		}
+
+		/* We can only read available data */
+		requested_size = min(requested_size, available_size);
+		if (!requested_size)
+			return -EAGAIN;
 	}
 
+	/* Size to end of buffer */
+	ptr->async_size = min(requested_size, desc->size - ptr->idx);
+
+	*raw_size_available = ptr->async_size;
+
+	/* Convert index to address in the buffer */
+	*buff = (void *)(desc->buff + ptr->idx);
+
+	ptr->async_started = true;
+
 	return ret;
 }
 
-/**
- * @brief Read data from the buffer
- * @param desc - Circular buffer reference
- * @param data - Buffer where to data is copied from the circular buffer
- * @param nb_elements - Number of elements to be copied
- * @return
- *  - \ref SUCCESS : On success
- *  - \ref FAILURE : If not enaught data available in the buffer
- */
-int32_t cb_read(struct circular_buffer *desc, void *data,
-		uint32_t nb_elements)
+static int32_t cb_end_async_operation(struct circular_buffer *desc,
+				      bool is_read)
 {
-	uint32_t	to_read_size;
-	uint32_t	size_to_end;
+	struct cb_ptr	*ptr;
+	uint32_t	new_val;
 
-	if (!desc || !nb_elements)
-		return FAILURE;
 
-	/* Check for overflow */
-	if (nb_elements > desc->nb_elem)
+	if (!desc)
+		return -EINVAL;
+
+	/* Select if read or write index will be updated */
+	ptr = is_read ? &desc->read : &desc->write;
+
+	/* Transaction not started */
+	if (!ptr->async_started)
 		return FAILURE;
-	desc->nb_elem -= nb_elements;
-
-	to_read_size = nb_elements * desc->elem_size;
-	size_to_end = desc->size - (desc->read_ptr - desc->buff);
-	if (to_read_size > size_to_end) {
-		memcpy(data, desc->read_ptr, size_to_end);
-		to_read_size -= size_to_end;
-		memcpy((int8_t *)data + size_to_end, desc->buff, to_read_size);
-		desc->read_ptr = desc->buff + to_read_size;
-	} else {
-		memcpy(data, desc->read_ptr, to_read_size);
-		desc->read_ptr += to_read_size;
+
+	/* Update pointer value */
+	new_val = ptr->idx + ptr->async_size;
+	if (new_val >= desc->size) {
+		ptr->spin_count++;
+		new_val %= desc->size;
+	}
+	ptr->idx = new_val;
+	ptr->async_size = 0;
+	ptr->async_started = false;
+
+	return SUCCESS;
+}
+
+static int32_t cb_operation(struct circular_buffer *desc,
+			    void *data, uint32_t size,
+			    bool is_read)
+{
+	uint8_t		*buff;
+	uint32_t	available_size;
+	int32_t		ret;
+	uint32_t	i;
+	bool		sticky_overrun;
+
+	if (!desc || !data || !size)
+		return -EINVAL;
+
+	sticky_overrun = 0;
+	i = 0;
+	while (i < size) {
+		do {
+			ret = cb_prepare_async_operation(desc, size - i,
+							 (void **)&buff,
+							 &available_size,
+							 is_read);
+		} while (ret == -EBUSY || ret == -EAGAIN);
```

Comment:
Blocking while data or space is not available. Is this intended?

Reply:
I thought a read operation should be blocking. So yes is intended.